R/read_csv.R

```diff
@@ -0,0 +1,239 @@
+#' Check that the sampling information from two CSV files matches.
+#' Will throw errors if the sampling informations dont match. If
+#' it returns, the sampling informations match.
+#'
+#' @param a the first sampling info to check
+#' @param b the second sampling info to check
+#'
+check_sampling_csv_info_matches <- function(a, b) {
+  if(a$stan_version_major != b$stan_version_major ||
+     a$stan_version_minor != b$stan_version_minor ||
+     a$stan_version_patch != b$stan_version_patch) {
+    stop("Supplied CSV files were not generated with the same version of Cmdstan!")
+  }
+  if(a$model != b$model) {
+    stop("Supplied CSV files were not generated wtih the same model!")
+  }
+  if(!all(a$params == b$params)) {
+    stop("Supplied CSV files have samples for different parameters!")
+  }
+  if(a$data_file != b$data_file) {
+    stop("Supplied CSV files have samples from chains run with non-matching data!")
+  }
+  for(name in names(a)) {
+    if(startsWith(name, "sample_")) {
+      if (is.null(b[[name]]) ||
+          a[[name]] != b[[name]]) {
+        stop("Supplied CSV files do not match in all sampling settings!")
+      }
+    }
+  }
+}
+
+#' Reads the sampling arguments and the diagonal of the
+#' inverse mass matrix from the comments in a CSV file.
+#'
+#' @param csv_file A CSV file containing results from sampling
+#' @return A list containing all sampling parameters and the
+#' diagonal of the inverse mass matrix
+#'
+read_sample_info_csv <- function(csv_file) {
+  param_names_read <- FALSE
+  sampling_params_read <- FALSE
+  diagonal_matrix_next <- FALSE
+  arg_prefix <- ""
+  csv_file_info = list()
+  con  <- file(csv_file, open = "r")
+  while (length(line <- readLines(con, n = 1, warn = FALSE)) > 0) {
+    if (!startsWith(line, "#")) {
+      if(!param_names_read) {
+        param_names_read <- TRUE
+        all_names <- strsplit(line, ",")[[1]]
+        csv_file_info[["sampler_params"]] <- c()
+        csv_file_info[["model_params"]] <- c()
+        for(x in all_names) {
+          if(endsWith(x, "__") && x != "lp__"){
+            csv_file_info[["sampler_params"]] <- c(csv_file_info[["sampler_params"]], x)
+          } else {
+            csv_file_info[["model_params"]] <- c(csv_file_info[["model_params"]], x)
+          }
+        }
+        next;
+      } else {
+        break;
+      }
+    }
+    # no more sampling settings follow the list of parameters
+    if(!param_names_read) {
+      tmp <- gsub("#", "", line, fixed = TRUE)
+      tmp <- gsub("(Default)", "", tmp, fixed = TRUE)
+      key_val <- grep("=", tmp, fixed = TRUE, value = TRUE)
+      if(length(key_val) == 0) {
+        arg_name <- trimws(tmp)
+        if(arg_prefix != "" && !sampling_params_read) {
+          arg_prefix <- paste(arg_prefix, arg_name, sep = "_")
```

Comment:
For now this builds a longer name like "sample_adapt_delta". Should it rather be ["sample"]["adapt"]["delta"]?

Reply:
maybe ["sample"]["adapt_delta"]? That is, `sample$adapt_delta`?

Reply:
Not a huge deal but really what would be great is to make the names match the names used in CmdStanArgs and SampleArgs when possible. What do you think?

Reply:
This function now returns the following

```R
list(
    model_name = csv_file_info$model,
    method = "sample",
    num_iter = num_iter,
    id = csv_file_info$id,
    num_warmup = csv_file_info$num_warmup,
    num_samples = csv_file_info$num_samples,
    save_warmup = csv_file_info$save_warmup,
    thin =  csv_file_info$thin,
    max_depth = csv_file_info$max_depth,
    adapt_engaged = csv_file_info$engaged,
    adapt_delta = csv_file_info$delta,
    stepsize = csv_file_info$stepsize,
    init_buffer = csv_file_info$init_buffer,
    term_buffer = csv_file_info$term_buffer,
    window = csv_file_info$window,
    model_params = csv_file_info$model_params,
    sampler_params = csv_file_info$sampler_params,
    inverse_mass_matrix = csv_file_info$inverse_mass_matrix
  )
```